assembly/src/assembler/module_graph/callgraph.rs

```diff
@@ -0,0 +1,399 @@
+            }
+            Err(CycleError(in_cycle))
+        } else {
+            assert!(!expect_cycle, "we expected a cycle to be found, but one was not identified");
+            Ok(output)
+        }
+    }
+
+    /// Gets a new graph which is a subgraph of `self` containing all of the nodes reachable from
+    /// `root`, and nothing else.
+    pub fn subgraph(&self, root: GlobalProcedureIndex) -> Self {
+        let mut worklist = VecDeque::from_iter([root]);
+        let mut graph = Self::default();
+        let mut visited = BTreeSet::default();
+
+        while let Some(gid) = worklist.pop_front() {
+            if !visited.insert(gid) {
+                continue;
+            }
+
+            let new_successors = graph.get_or_insert_node(gid);
+            let prev_successors = self.out_edges(gid);
+            worklist.extend(prev_successors.iter().cloned());
+            new_successors.extend_from_slice(prev_successors);
+        }
+
+        graph
+    }
+
+    /// Constructs the topological ordering of nodes in the call graph, for which `caller` is an
+    /// ancestor.
+    ///
+    /// # Errors
+    /// Returns an error if a cycle is detected in the graph.
+    pub fn toposort_caller(
+        &self,
+        caller: GlobalProcedureIndex,
+    ) -> Result<Vec<GlobalProcedureIndex>, CycleError> {
+        let mut output = Vec::with_capacity(self.nodes.len());
+
+        // Build a subgraph of `self` containing only those nodes reachable from `caller`
+        let mut graph = self.subgraph(caller);
+
+        // Remove all predecessor edges to `caller`
+        graph.nodes.iter_mut().for_each(|(_pred, out_edges)| {
+            out_edges.retain(|n| *n != caller);
+        });
+
+        let mut roots = VecDeque::from_iter([caller]);
+        let mut successors = Vec::with_capacity(4);
+        while let Some(id) = roots.pop_front() {
+            output.push(id);
+            successors.clear();
+            successors.extend(graph.nodes[&id].iter().copied());
+            for mid in successors.drain(..) {
+                graph.remove_edge(id, mid);
+                if graph.num_predecessors(mid) == 0 {
+                    roots.push_back(mid);
+                }
+            }
+        }
+
+        let has_cycle = graph
+            .nodes
+            .iter()
+            .any(|(n, out_edges)| output.contains(n) && !out_edges.is_empty());
+        if has_cycle {
+            let mut in_cycle = BTreeSet::default();
+            for (n, edges) in graph.nodes.iter() {
+                if edges.is_empty() {
+                    continue;
+                }
+                in_cycle.insert(*n);
+                in_cycle.extend(edges.as_slice());
+            }
+            Err(CycleError(in_cycle))
+        } else {
+            Ok(output)
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    use crate::{
+        assembler::{GlobalProcedureIndex, ModuleIndex},
```

Comment:
nit: I think `GlobalProcedureIndex` and `ModuleIndex` are covered by the `super::*` import (or at least i get a warning about their imports).